### PR TITLE
Remote window component desktop

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers a continuation of this work: https://github.com/keybase/client/pull/1384

Now all you have to do is:
```js
<RemoteComponent
  component='pinentry'
  somePinentryProp='foo'/>
```

No change is required for the pinentry component itself!

To get this working we implement a `RemoteStore` that implements redux's `Store` interface. Then we can just use `Provider`/`ConnectedComponent` as before. This has the downside that if a component needs some part of the state it has to wait for the state to be serialized and sent over ipc before it can interact with it (and on every state transition).

I don't think that will be that big of a problem (at least in the beginning) since the simplicity and consistency outweigh the slowness.

There is also a workaround which is to make a wrapping component that takes the things out of the store it needs and passes it as props to the `RemoteComponent`. That'll make sure all state access happens locally.

This doesn't implement the existing pinentry flow in terms of this. I'll make a separate PR for that

### CODE REVIEW

**Comment by MarcoPolo** [comment]
I know I need to change the require ipc's to ipcMain and ipcRenderer, but that's an easy change and can be done in parallel to review

File: react-native/react/native/remote-component-loader.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      action(a => this.dispatch(a))
+    } else {
+      ipc.send('dispatchAction', action)
+    }
+  }
+
+  subscribe (listener) {
+    this.listeners.push(listener)
+    return listener => {
+      this.listeners = this.listerns.filter(l => l !== listener)
```
Comment (cjb) [inline react-native/react/native/remote-component-loader.js]: is `listerns` a typo?

**Comment by chrisnojima** [comment]
typo aside seems like a good move forward :+1:

File: react-native/react/native/remote-component-loader.js (hunk 1 above)
Comment (MarcoPolo) [inline react-native/react/native/remote-component-loader.js]: thank you, that would have been a very annoying bug...

**Comment by cjb** [comment]
:thumbsup:!

File: desktop/renderer/index.desktop.js [hunk 2]
```diff
@@ -8,6 +8,11 @@ import Nav from '../../react-native/react/nav'
 import injectTapEventPlugin from 'react-tap-event-plugin'
 import {isDev} from '../../react-native/react/constants/platform'
 
+// For Remote Components
+import remote from 'remote'
+const ipc = remote.require('ipc')
+// import ipc from 'ipc'
```
Comment (cjb) [inline desktop/renderer/index.desktop.js]: remove this line maybe?

Comment (MarcoPolo) [inline desktop/renderer/index.desktop.js]: yep! I'm going to add another commit that will change the ipc to ipcMain/ipcRenderer

Comment (cjb) [inline desktop/renderer/index.desktop.js]: Cool.  As a heads up, for me s/ipc/ipcMain/g worked great, but s/ipc/ipcRenderer/g in my component gave me a weird crash, haven't figured it out why yet.

**Pull request merged**